_posts/en/newsletters/2019-03-05-newsletter.md

```diff
@@ -0,0 +1,436 @@
+    concepts such as `OP_CODESEPARATOR`, `FindAndDelete()`, and time
+    warp attacks, so we've also included an appendix to this newsletter
+    that provides additional background on these subjects.
+
+    - **Prevent use of `OP_CODESEPARATOR` and `FindAndDelete()` in
+      legacy transactions:** nobody is known to be using these two
+      features of Bitcoin in legacy (non-segwit) Bitcoin transactions,
+      but an attacker can abuse them to significantly increase the
+      amount of computational work necessary to verify a non-standard
+      transaction, creating blocks that could take half an hour or
+      longer to verify.  Most readers will probably never have heard
+      about either of these features because neither one of them enables
+      any known useful behavior that can't be accomplished in some other
+      way, but anyone who does still need `OP_CODESEPARATOR` can use
+      the [BIP143][] segwit version that was implemented to avoid the
+      computational blowup.
+
+{% comment %}<!--
+      ## How long until all bitcoins are released by a timewarp
+      ## There are 26 retarget periods per year, about 100 years of subsidy left: 2600
+      ## Calculate in days
+      In [7]: x = 0
+         ...: for i in range(2600):
+         ...:     x += 14*(0.25**i)
+         ...: print(x)
+         ...:
+         18.666666666666668 -->{% endcomment %}
+
+    - **Fix the time warp attack:** this attack allows miners controlling
+      a majority of hashrate to decrease mining difficulty even when
+      total network hashrate is steady or increasing, allowing them to
+      produce blocks faster than targeted by the protocol.  The increase
+      in block production would accelerate release of Bitcoin's block
+      subsidy, potentially releasing all remaining bitcoins within three
+      weeks of the attack starting.  However, the setup for the attack
+      would be publicly visible for at least a week before it had any
+      effect, so fixing it has not had a high priority in the absence of
+      a cartel of miners attempting it.  The proposed soft fork fixes
+      the problem by requiring the first block in a new difficulty
+      period have a timestamp no more than 600 seconds earlier than the
+      last block in the previous period.  See also [Newsletter FIXME][]
+      where we mention a mailing list discussion about the topic.
+
+    - **Encourage retargeting blocks to a 600 second average:**
+      probability theory, backed by simulation, [shows][sipa gist that
+      defies my intuition so hard it hurts] that Bitcoin's
+      implementation causes the difficulty adjustment algorithm to
+      attempt to keep the average time between blocks at about 600.6
+      seconds instead of the intended 600.0 seconds.  This can be fixed
+      by always setting the timestamp of the first block in a new
+      difficulty period to 600 seconds before its previous block.  So,
```

Comment:
(Comment mainly for other reviewers, rather than @jnewbery)

This was interesting to think about but, after reflection, I do think the existing text is correct.  The goal of the idea is to make the retarget formula only consider 2014 blocks of the 2016 period.  We get one block not considered for free because of the off-by-one error.  We get the second block by moving time back one average block interval before it was mined (600 seconds).  Matt's suggested code implements this as described: `https://github.com/bitcoin/bitcoin/pull/15481/files#diff-9651347c8e00bed3ddc7631de569406dR656`

Reply:
I think I misread the BIP.  I don't think it addresses this "problem" of 600.6 second target intervals at all, so I'm deleting all the text related to it.